File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = set()
        max_len=0
        l=0
        for r in range(len(s)):
            x = s[r]
            if x in seen:
                while x in seen:
                    seen.remove(s[l])
                    l+=1
            seen.add(x)
            cl = r-l+1
            max_len=max(max_len,cl)
        return max_len
                

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] used = new int[128];
        int maxLen = 0;
        int st = 0;
        int end = 0;
        while(end<s.length() && st<=end){
            used[s.charAt(end)]++;
            while(used[s.charAt(end)]>1){
                used[s.charAt(st)]--;
                st++;
            }
            maxLen = Math.max(maxLen, end-st+1);
            end++;
        }
        return maxLen;
    }
}